src/utils/usePromise.js

```diff
@@ -0,0 +1,81 @@
+/* eslint react/prop-types: 0 */
+// @flow
+import * as React from 'react';
+
+function resolvePromise(promise) {
+  if (typeof promise === 'function') {
+    return promise();
+  }
+
+  return promise;
+}
+
+const states = {
+  pending: 'pending',
+  rejected: 'rejected',
+  resolved: 'resolved',
+  initial: 'initial',
+};
+
+const STATE_PENDING = {
+  error: undefined,
+  result: undefined,
+  state: states.pending,
+};
+
+function reducer(state, action) {
+  switch (action.type) {
```

Comment:
I think this could be modified to preserve the error/result states, and then we don't need to do the work to track those in useAsync.